### PR TITLE
fill with a solid color if the sky is not stretched

### PR DESCRIPTION
**Pull request by rfomin**
I still like this idea. I think it looks better than tiled sky.

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
This looks familiar... 🤔 😄

**Comment by rfomin** [comment]
> This looks familiar... :thinking: :smile:

Yeah, I used you patch :smile: Thinking about adding fade to color.

File: Source/r_plane.c [hunk 1]
```diff
@@ -389,9 +396,30 @@ static void do_draw_plane(visplane_t *pl)
 	  {
 	    dc_texturemid = skytexturemid;    // Default y-offset
 	    texture = skytexture;             // Default texture
+	    dc_texheight = textureheight[texture]>>FRACBITS;
 	    flip = 0;                         // Doom flips it
 	  }
 
+	ttop = centery - viewheight/2 + (128 - dc_texheight) * (hires ? 2 : 1);
```
Comment (rfomin) [inline Source/r_plane.c:403]: There is probably a better way to support tall skies?

File: Source/r_plane.c [hunk 2]
```diff
@@ -375,6 +379,9 @@ static void do_draw_plane(visplane_t *pl)
 
 	    // Vertical offset allows careful sky positioning.
 
+	    if (dc_texheight >= 200)
```
Comment (fabiangreffrath) [inline Source/r_plane.c:382]: I'd say, if we decide to second-guess `dc_texturemid` for tall skies, we should completely ignore `s->rowoffset` and just let it start at the top, as we do for regular sky textures.

**Comment by fabiangreffrath** [comment]
Let's make this independent of the "Stretch Short Skies" setting: If there is a gap between the top of the sky texture and the top rendered screen line, apply a solid color. This covers both the unstretched short skies and the "misaligned" tall skies (as in Eternity) cases.

**Comment by fabiangreffrath** [comment]
I like this logic:

- We don't want to *move* MBF skies, because the author already made that choice.
- We don't want to *stretch* tall skies, because they are already tall.

This is the situation we currently have, and I'd like to keep it. However, this will leave us with either tiled sky textures or gaps above them. If we fill these gaps with solid color (or a gradient), that'd be fine with me.

**Comment by rfomin** [comment]
I agree, so I just have to figure out how to do it :smile: Do I need new` r_draw.c` function for it?

**Comment by fabiangreffrath** [comment]
No further function necessary. It's basically what you are already doing with the `ttop` variable, i.e. check if the sky texture top line has a positive value - which would be the gap. Then draw the solid color from y = 0 down to ttop.

However, I am not sure about the current `ttop` calculation...

**Comment by fabiangreffrath** [comment]
Sorry, I must have misunderstood your question. Also, I was at least one commit behind your PR on my local branch.

So, indeed, if you want to do a proper color gradient, you will need a new column drawing function. Additionally you will need information how the top row colors of the sky texture transform to black. The colormaps table could help here.

**Comment by rfomin** [comment]
I ended up copying EE's solution, I think it looks good.
Doom 2 map15:
doom02
Eviternity map11:
doom03

File: Source/r_draw.c [hunk 3]
```diff
@@ -82,7 +82,8 @@ int     dc_yh;
 fixed_t dc_iscale; 
 fixed_t dc_texturemid;
 int     dc_texheight;    // killough
-byte    *dc_source;      // first pixel in a column (possibly virtual) 
+byte    *dc_source;      // first pixel in a column (possibly virtual)
```
Comment (fabiangreffrath) [inline Source/r_draw.c:85]: There must be some unrelated whitespace change.

File: Source/r_draw.c [hunk 4]
```diff
@@ -270,6 +271,133 @@ void R_DrawTLColumn (void)
   }
 } 
 
+//
```
Comment (fabiangreffrath) [inline Source/r_draw.c:274]: Please quote the sources here, e.g. `eternity-engine/source/r_draw.c:L200-400`.

File: Source/r_plane.c [hunk 5]
```diff
@@ -375,6 +378,9 @@ static void do_draw_plane(visplane_t *pl)
...
+	      dc_texturemid = s->rowoffset + 200*FRACUNIT;
+	    else
 	    dc_texturemid = s->rowoffset - 28*FRACUNIT;
```
Comment (fabiangreffrath) [inline Source/r_plane.c:384]: Do we still move the sky texture against the map author's choice?

File: Source/r_sky.c [hunk 6]
```diff
@@ -32,6 +32,10 @@
 
 #include "r_sky.h"
 #include "r_state.h" // [FG] textureheight[]
+#include "r_data.h"
+#include "w_wad.h"
+
+extern int *texturewidth;
```
Comment (fabiangreffrath) [inline Source/r_sky.c:38]: I included `r_state.h` only for `textureheight[]`, so if you like please  add `texturewidth[]` to `r_state.h`.

File: Source/r_sky.c [hunk 7]
```diff
@@ -67,6 +71,73 @@ void R_InitSkyMap (void)
...
+    g += pal[3 * c[0] + 1];
+    b += pal[3 * c[0] + 2];
+  }
+
+  r /= width;
+  g /= width;
+  b /= width;
+
+  // Get 1/3 for empiric reasons
+  ret = V_GetPaletteIndex(pal, r/3, g/3, b/3);
```
Comment (fabiangreffrath) [inline Source/r_sky.c:133]: This returns a color that is very much darker than the average sky top color. Is this intended?

File: Source/r_sky.c [hunk 8]
```diff
@@ -67,6 +71,73 @@ void R_InitSkyMap (void)
...
+    if (diff == 0)
+    {
+      break;
+    }
+  }
+
+  return best;
+}
+
+byte R_SkyBlendColor(int tex)
```
Comment (fabiangreffrath) [inline Source/r_sky.c:103]: I am having my problems with this function, because it only caches the return value for the last requested sky texture. In theory you may have an unlimited number of MBF skies in a map, all visible at the same time. This means that this function will be called over and over again during each rendered frame, which will be a performance desaster. 

I am playing the thought of creating a dynamic texture->color lookup table and reallocate it on demand, so that each texture has its color determined only once. What do you think of this?

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]

File: Source/r_sky.c (hunk 7 above)
Comment (rfomin) [reply Source/r_sky.c:133]: Yes this is intended I copied it form EE. Darker color looks better, for example on Doom 2 map15.

File: Source/r_sky.c (hunk 8 above)
Comment (rfomin) [reply Source/r_sky.c:103]: I agree, in fact EE does something similar: https://github.com/team-eternity/eternity/blob/85d5105207edf856d664932c428099dffb228b69/source/r_sky.cpp#L198

File: Source/r_draw.c (hunk 4 above)
Comment (rfomin) [reply Source/r_draw.c:274]: Fixed in https://github.com/fabiangreffrath/woof/pull/483/commits/ee455c291efa9afb2d6fb3b5e3f757e87a001de5

File: Source/r_sky.c (hunk 6 above)
Comment (rfomin) [reply Source/r_sky.c]: Fixed in https://github.com/fabiangreffrath/woof/pull/483/commits/ee455c291efa9afb2d6fb3b5e3f757e87a001de5

File: Source/r_plane.c (hunk 5 above)
Comment (rfomin) [reply Source/r_plane.c]: Fixed in https://github.com/fabiangreffrath/woof/pull/483/commits/f04ed9607e3613334cd656ccb03674ba0aec5989

File: Source/r_plane.c [hunk 9]
```diff
@@ -49,6 +49,7 @@
 #include "r_things.h"
 #include "r_sky.h"
 #include "r_plane.h"
+#include "v_video.h"
```
Comment (fabiangreffrath) [inline Source/r_plane.c:52]: What do we need this for?

**Comment by fabiangreffrath** [comment]
You could even pre-fill the table in `R_InitSkyMap()` with the regular and all MBF sky textures.

File: Source/r_sky.c [hunk 10]
```diff
@@ -67,6 +69,114 @@ void R_InitSkyMap (void)
   skytexturemid = 100*FRACUNIT;
 }
 
+static byte V_GetPaletteIndex(byte *palette, int r, int g, int b)
```
Comment (fabiangreffrath) [inline Source/r_sky.c:72]: This function taken from Chocolate Doom's `i_video.c` (I think). Could you please document this as before for `R_DrawSkySpan()`?

Comment (rfomin) [reply Source/r_sky.c:73]: Fixed in https://github.com/fabiangreffrath/woof/pull/483/commits/7dc98d16ed4873ace4264305216c4135343ec9c4

**Comment by rfomin** [comment]
> You could even pre-fill the table in `R_InitSkyMap()` with the regular and all MBF sky textures.

For regular sky, this could be done easily, but not for MBF? We get `texturenum` from the sidedef in `do_draw_plane`, right? Anyway, I think there is very little performance gain from this.

**Comment by fabiangreffrath** [comment]
> For regular sky, this could be done easily, but not for MBF? We get `texturenum` from the sidedef in `do_draw_plane`, right? Anyway, I think there is very little performance gain from this.

We could even do this very early when spawning the map's specials:

https://github.com/fabiangreffrath/woof/blob/96cc3704893e52fc4b49cc63f5dfee1a7a73347c/Source/p_spec.c#L2415-L2428

File: Source/p_spec.c [hunk 11]
```diff
@@ -2505,7 +2506,15 @@ void P_SpawnSpecials (void)
       case 271:   // Regular sky
       case 272:   // Same, only flipped
         for (s = -1; (s = P_FindSectorFromLineTag(lines+i,s)) >= 0;)
+        {
+          // Texture comes from upper texture of reference sidedef
```
Comment (fabiangreffrath) [inline Source/p_spec.c:2510]: Does this have to be in the `for()` loop?

Comment (fabiangreffrath) [reply Source/p_spec.c:2510]: Well, in the lines above this is separated. Also, not fixing inventory bugs does not mean to introduce similar new ones. 😉

Comment (rfomin) [reply Source/p_spec.c]: Do you mean this: https://github.com/fabiangreffrath/woof/pull/483/commits/a9682012aefb4cbba0c695ccc6358b062cdbdf0f ?

Comment (fabiangreffrath) [reply Source/p_spec.c]: Beautiful, thank you!

**Review by fabiangreffrath — APPROVED** [review]
Very nice! Ready to merge?

**Comment by rfomin** [comment]
> Very nice! Ready to merge?

Yes, I think so.
